File: backend/app/services/categorizer.py
import os
import logging

logger = logging.getLogger(__name__)

try:
    import joblib
    HAS_JOBLIB = True
except ImportError:
    HAS_JOBLIB = False

MODEL_PATH = os.path.join(os.path.dirname(__file__), "..", "ml", "model.pkl")
_model = None

def load_model():
    global _model
    if HAS_JOBLIB and os.path.exists(MODEL_PATH):
        try:
            _model = joblib.load(MODEL_PATH)
            logger.info("ML categorization model loaded successfully.")
        except Exception as e:
            logger.error(f"Failed to load ML model: {e}")
            _model = None
    else:
        logger.warning(f"ML model not found at {MODEL_PATH}. Falling back to basic categorization.")

def basic_categorize(merchant: str) -> str:
    merchant = merchant.lower()
    if any(x in merchant for x in ['zomato', 'swiggy', 'mcdonalds', 'kfc', 'pizza', 'restaurant']):
        return 'Food & Dining'
    elif any(x in merchant for x in ['netflix', 'spotify', 'amazon prime', 'hotstar', 'movie', 'cinema']):
        return 'Entertainment'
    elif any(x in merchant for x in ['amazon', 'flipkart', 'myntra', 'shopping', 'store', 'mall']):
        return 'Shopping'
    elif any(x in merchant for x in ['uber', 'ola', 'irctc', 'petrol', 'flight', 'railway', 'transport']):
        return 'Transport'
    elif any(x in merchant for x in ['jio', 'airtel', 'electricity', 'water', 'gas', 'recharge']):
        return 'Utilities'
    elif any(x in merchant for x in ['rent', 'maintenance', 'housing', 'broker']):
        return 'Housing'
    elif any(x in merchant for x in ['gym', 'pharmacy', 'hospital', 'doctor', 'med']):
        return 'Health & Fitness'
    return 'Miscellaneous'

def predict(merchant_name: str) -> str:
    if _model is None:
        return basic_categorize(merchant_name)
        
    try:
        probs = _model.predict_proba([merchant_name])[0]
        max_prob = max(probs)
        if max_prob < 0.4:
            return 'Miscellaneous'
        return _model.classes_[probs.argmax()]
    except Exception as e:
        logger.error(f"Prediction error for {merchant_name}: {e}")
        return basic_categorize(merchant_name)

def predict_batch(merchants: list[str]) -> list[str]:
    if _model is None:
        return [basic_categorize(m) for m in merchants]
        
    try:
        probs = _model.predict_proba(merchants)
        results = []
        for prob in probs:
            if max(prob) < 0.4:
                results.append('Miscellaneous')
            else:
                results.append(_model.classes_[prob.argmax()])
        return results
    except Exception as e:
        logger.error(f"Batch prediction error: {e}")
        return [basic_categorize(m) for m in merchants]

File: frontend/src/lib/categorizer.ts
export const CATEGORIES: string[] = [
  'Food & Dining',
  'Entertainment',
  'Shopping',
  'Transport',
  'Utilities',
  'Housing',
  'Health & Fitness',
  'Miscellaneous'
];

const EXACT_LOOKUP = new Map<string, string>([
  ['Zomato', 'Food & Dining'], ['Swiggy', 'Food & Dining'], ['McDonalds', 'Food & Dining'], ['Dominos', 'Food & Dining'], ['Pizza Hut', 'Food & Dining'], ['Starbucks', 'Food & Dining'], ['KFC', 'Food & Dining'], ['Subway', 'Food & Dining'], ['Burger King', 'Food & Dining'], ['Haldirams', 'Food & Dining'], ['Barbeque Nation', 'Food & Dining'], ['Chai Point', 'Food & Dining'], ['Dunkin', 'Food & Dining'], ['Baskin Robbins', 'Food & Dining'],
  ['Netflix', 'Entertainment'], ['Spotify', 'Entertainment'], ['Amazon Prime', 'Entertainment'], ['Hotstar', 'Entertainment'], ['YouTube Premium', 'Entertainment'], ['BookMyShow', 'Entertainment'], ['PVR Cinemas', 'Entertainment'], ['INOX', 'Entertainment'], ['Sony LIV', 'Entertainment'], ['Apple Music', 'Entertainment'],
  ['Amazon', 'Shopping'], ['Flipkart', 'Shopping'], ['Myntra', 'Shopping'], ['Ajio', 'Shopping'], ['Meesho', 'Shopping'], ['Nykaa', 'Shopping'], ['Decathlon', 'Shopping'], ['IKEA', 'Shopping'], ['H&M', 'Shopping'], ['Zara', 'Shopping'], ['Nike', 'Shopping'], ['Reliance Digital', 'Shopping'], ['Croma', 'Shopping'],
  ['Uber', 'Transport'], ['Ola', 'Transport'], ['Rapido', 'Transport'], ['Metro', 'Transport'], ['IRCTC', 'Transport'], ['MakeMyTrip', 'Transport'], ['Goibibo', 'Transport'], ['Yulu', 'Transport'], ['Bounce', 'Transport'], ['Shell Petrol', 'Transport'], ['HP Petrol', 'Transport'], ['Indian Oil', 'Transport'],
  ['Jio', 'Utilities'], ['Airtel', 'Utilities'], ['Vi', 'Utilities'], ['BSNL', 'Utilities'], ['Electricity Board', 'Utilities'], ['Water Board', 'Utilities'], ['Gas', 'Utilities'], ['Tata Power', 'Utilities'], ['Adani Electricity', 'Utilities'], ['Broadband', 'Utilities'],
  ['Rent', 'Housing'], ['Housing Society', 'Housing'], ['Maintenance', 'Housing'], ['Property Tax', 'Housing'], ['Home Loan EMI', 'Housing'], ['NoBroker', 'Housing'],
  ['Gym', 'Health & Fitness'], ['Cult Fit', 'Health & Fitness'], ['Apollo Pharmacy', 'Health & Fitness'], ['Netmeds', 'Health & Fitness'], ['PharmEasy', 'Health & Fitness'], ['1mg', 'Health & Fitness'], ['Practo', 'Health & Fitness'], ['Lenskart', 'Health & Fitness'],
  ['ATM Withdrawal', 'Miscellaneous'], ['Bank Transfer', 'Miscellaneous'], ['Cash', 'Miscellaneous'], ['UPI', 'Miscellaneous']
]);

const EXACT_LOOKUP_LOWER = new Map<string, string>();
EXACT_LOOKUP.forEach((val, key) => {
  EXACT_LOOKUP_LOWER.set(key.toLowerCase(), val);
});

function basicCategorize(merchant: string): string {
  const lower = merchant.toLowerCase();
  if (['zomato', 'swiggy', 'mcdonalds', 'kfc', 'pizza', 'restaurant'].some(x => lower.includes(x))) {
    return 'Food & Dining';
  } else if (['netflix', 'spotify', 'amazon prime', 'hotstar', 'movie', 'cinema'].some(x => lower.includes(x))) {
    return 'Entertainment';
  } else if (['amazon', 'flipkart', 'myntra', 'shopping', 'store', 'mall'].some(x => lower.includes(x))) {
    return 'Shopping';
  } else if (['uber', 'ola', 'irctc', 'petrol', 'flight', 'railway', 'transport'].some(x => lower.includes(x))) {
    return 'Transport';
  } else if (['jio', 'airtel', 'electricity', 'water', 'gas', 'recharge'].some(x => lower.includes(x))) {
    return 'Utilities';
  } else if (['rent', 'maintenance', 'housing', 'broker'].some(x => lower.includes(x))) {
    return 'Housing';
  } else if (['gym', 'pharmacy', 'hospital', 'doctor', 'med'].some(x => lower.includes(x))) {
    return 'Health & Fitness';
  }
  return 'Miscellaneous';
}

export function categorize(merchantName: string): string {
  if (!merchantName) return 'Miscellaneous';
  const exactMatch = EXACT_LOOKUP_LOWER.get(merchantName.toLowerCase());
  if (exactMatch) {
    return exactMatch;
  }
  return basicCategorize(merchantName);
}

export function categorizeBatch(merchants: string[]): string[] {
  return merchants.map(categorize);
}
